crates/apollo-router-core/src/query_planner/caching_query_planner.rs

```diff
@@ -37,25 +37,75 @@ impl<T: QueryPlanner> QueryPlanner for CachingQueryPlanner<T> {
         }
 
-        let value = self
-            .delegate
-            .get(query.clone(), operation.clone(), options.clone())
-            .await;
-        self.cached
-            .lock()
-            .await
-            .insert((query, operation, options), value.clone());
-        value
+        // Holding a lock across the delegated get is a bad idea because
+        // the delegate get() calls into v8 for processing of the plan.
+        // This would block all other get() requests for a potentially
+        // long time.
+        // Alternatively, if we don't hold the lock, there is a risk
+        // that we will do the work multiple times. This is also
+        // sub-optimal.
+
+        // To work around this, we keep a list of keys we are currently
+        // processing in the delegate. If we try to get a key on this
+        // list, we block and wait for it to complete and then retry.
+        //
+        // This is more complex than either of the two simple
+        // alternatives but succeeds in providing a mechanism where each
+        // client only waits for uncached QueryPlans they are going to
+        // use AND avoids generating the plan multiple times.
+
+        drop(locked_cache);
+
+        let mut locked_wait_map = self.wait_map.lock().await;
+
+        match locked_wait_map.get_mut(&key) {
+            Some(waiter) => {
+                // Register interest in key
+                let mut receiver = waiter.subscribe();
+                drop(locked_wait_map);
+                let msg = receiver.recv().await?;
+                assert_eq!(msg, key);
+                let mut locked_cache = self.cached.lock().await;
+                if let Some(value) = locked_cache.get(&key).cloned() {
+                    return value;
+                }
+                unreachable!();
+            }
+            None => {
+                let (tx, _rx) = broadcast::channel(10);
+                locked_wait_map.insert(key.clone(), tx.clone());
+                drop(locked_wait_map);
+                // This is the potentially high duration operation
+                let value = self
+                    .delegate
+                    .get(key.0.clone(), key.1.clone(), key.2.clone())
+                    .await;
+                // Update our cache
+                let mut locked_cache = self.cached.lock().await;
+                locked_cache.put(key.clone(), value.clone());
+                // Update our wait list
+                let mut locked_wait_map = self.wait_map.lock().await;
+                locked_wait_map.remove(&key);
+                // Let our waiters know
+                tokio::task::spawn_blocking(move || tx.send(key)).await??;
+                value
+            }
+        }
+    }
+
+    async fn get_hot_keys(&self) -> Vec<QueryKey> {
+        let locked_cache = self.cached.lock().await;
+        let mut results = vec![];
+        for (key, _value) in locked_cache.iter().take(self.query_cache_limit / 5) {
```

Comment:
`let results = locked_cache.keys().take(self.query_cache_limit / 5).cloned().collect::<Vec<_>>()` maybe?

Reply:
I already commented on that plus my solution work.... 

wait... where is my comment :thinking:

Reply:
Somehow it disappeared I don't know :woman_shrugging: Here, I recreated it: https://github.com/apollographql/router/pull/174#discussion_r756356657

(And no you can't use `.keys()` because it's not a hashmap and you shouldn't use `.cloned()` without `.keys()` because it would clone the values too. And you don't need to pass the type argument because it's inferred from the function signature)

Reply:
right, and it would not make sense to iterate on keys, since we want them in most recently used order. I guess `let results = locked_cache.iter()().take(self.query_cache_limit / 5).map(|(_k, v)| v.clone()).collect()` wuld work then

Reply:
huh and _now_ I see your other comment :sweat_smile:

Reply:
resolved in other comment.